Dodx 1511 show almost there message

jonverson opened the pull request:
# Description
Added in logic to show your almost there logic after the 4 PR and the timer is showing

# Test process
<!--
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.
-->

- If you go to the profile page with with at least 4 PR's and at least one of them has the timer you will get the almost there.


# Requirements to merge
<!--
Ensure your pull request meets all the requirements for merging. Place an `x` in each box to indicate that your pull request meets that requirement.
-->
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

MattIPv4 reviewed (changes requested):
I don't think this will work as expected if one of the first four PRs is marked as spam.

I think the correct fix here is to update scoring_pull_requests to only include maximum four waiting PRs. This should be a change to https://github.com/digitalocean/hacktoberfest/blob/030d3436c11464243fcde66b9ebc4180855de901/app/services/pull_request_service.rb#L49 to make that add one to the count for a waiting OR eligible PR.

jonverson commented:
Hey @MattIPv4 - I made some updates which is how I believe it should work. Let me know if I am misunderstanding anything. I updated the scoring_pull_requests method to include eligible and waiting pr's. Then in the timeline partial I will show the your almost there message if you have had 4 eligible or waiting PR's and you haven't won yet. If you have won it will show the you won message instead after the 4th PR.

jonverson commented:
@MattIPv4 - I updated the logic for showing the your almost there message and you won message. Moved this logic into the pr partial. This PR should also cove the bonus icons. It will show all icons that are eligible that are over the 4 PR's as bonus ones.

MattIPv4 commented on app/views/users/show/pr/_icon.html.erb:
```diff
@@ -1,3 +1,10 @@
+<% if valid_pr > 4 && pr.eligible?%>
```
Should this not be moved down to be inside the `<% elsif pr.eligible? %>` block, so if its a bonus PR, we show the bonus icon, else we show the eligible icon?

MattIPv4 commented on app/views/users/show/pr/_pr.html.erb:
```diff
@@ -1,25 +1,90 @@
-    <div class="timeline-wrapper">
-      <div class="pr-icon">
-        <%= render partial: 'users/show/pr/icon', locals: { pr: pr } %>
+<% if valid_pr == 4 && !show_congratulations %>
```
There seems to be a huge amount of duplicated content in this file now, shouldn't we only need the `winning-msg-cont` part inside the if?

MattIPv4 commented on app/views/users/show/_timeline.html.erb:
```diff
@@ -3,30 +3,14 @@
-          <%= render partial: 'users/show/pr/pr', locals: { pr: pr } %>
-        <% end %>
-
-        <% if show_timer %>
```
It looks like we're no longer using `show_timer` -- assuming we're not, we can remove the helper method definition

jonverson commented:
@MattIPv4 - I have those changes up for you on this branch.

MattIPv4 commented on app/views/users/show/_timeline.html.erb:
```diff
@@ -3,40 +3,45 @@
+          <% end %>
+        <% end %>
+
+        <% if scoring_pull_requests.length == 4 && !show_congratulations %>
```
Why are we using this logic, and not just the original `show_timer`? That was based on user state, which ensures we're showing the message when expected, whereas this adds more complex logic to achieve hopefully the same thing?

MattIPv4 commented on app/views/users/show/pr/_pr.html.erb:
```diff
@@ -1,25 +1,59 @@
-<tr>
```
I'm really confused by the changes to this file -- shouldn't this just match master, with like a one-line change to pass the new `pr_type` through to the icon partial?

MattIPv4 commented on app/views/users/show/_timeline.html.erb:
```diff
@@ -3,40 +3,45 @@
         <% if non_scoring_pull_requests.any? %>
           <% non_scoring_pull_requests.map do |pr| %>
-            <%= render partial: 'users/show/pr/pr', locals: { pr: pr } %>
+            <%= render partial: 'users/show/pr/pr', locals: { pr: pr, pr_type: 0, valid_pr: @valid_pr, show_congratulations: show_congratulations } %>
```
Why do we need to pass `valid_pr` and `show_congratulations` in? The messaging is being shown in this template, not the PR partial, and we're not using `valid_pr` anywhere?

MattIPv4 commented on app/views/users/show/_timeline.html.erb:
```diff
@@ -3,40 +3,45 @@
-        <% if show_congratulations %>
+        <% scoring_pull_requests.map.with_index do |pr| %>
+          <% if pr.eligible? || pr.waiting? %>
+            <%= render partial: 'users/show/pr/pr', locals: { pr: pr, valid_pr: @valid_pr +=1, show_congratulations: show_congratulations, pr_type: 1 }  %>
```
As per my other comment, I'm really confused why we're passing `valid_pr` and `show_congratulations`, neither seem to be doing anything and just adding confusing complexity?

MattIPv4 commented on app/views/users/edit.html.erb:
```diff
@@ -8,9 +8,18 @@
 
         <%= form_for :user, url: { action: 'update' }, method: :patch do |f| %>
-          <%= render partial: 'users/dropdowns', locals: { f: f, u: @current_user } %>
+          <div id="form-first">
```
I think it's best we revert these changes in this PR -- let's keep this PR focused on the timeline updates, I'll be fixing this up in #656

MattIPv4 commented on app/views/users/show/_timeline.html.erb:
```diff
@@ -3,40 +3,40 @@
-        <% end %>
-
-        <% if show_timer %>
+        <% if show_congratulations %>
```
Is there a reason why we're now showing this congrats message above the scoring PRs?

I think it'd make far more sense to keep it rendering between the scoring & non_scoring PRs, especially as once the user has "won" and the scoring PRs is a frozen array, they might not be in chronological order with the non_scoring PRs -- so having the message to split the timeline makes sense?

![image](https://user-images.githubusercontent.com/12371363/95654359-6903f900-0af7-11eb-810d-5690de8e3786.png)

MattIPv4 commented on app/views/users/show/_timeline.html.erb:
```diff
@@ -4,15 +4,15 @@
               <div class="winning-msg-cont">
                 <h2 class="winning-title">You're almost there!</h2>
-                <span class="winning-desc">🎉 You’ve submitted the four required PRs for the Hacktoberfest challenge! So long as your PRs successfully pass the review period, they’ll become valid — which means you’ll have officially completed this year’s challenge!</span>
+                <span class="winning-desc">🎉 You’ve submitted the four required PRs for the Hacktoberfest challenge! So long as your PRs successfully pass the review period, they’ll become valid in — which means you’ll have officially completed this year’s challenge!</span>
```
```suggestion
                <span class="winning-desc">🎉 You’ve submitted the four required PRs for the Hacktoberfest challenge! So long as your PRs successfully pass the review period, they’ll become valid — which means you’ll have officially completed this year’s challenge!</span>
```
Rm `in`

MattIPv4 commented on app/views/users/show/pr/_icon.html.erb:
```diff
@@ -13,6 +13,11 @@
     <div class="unaccepted-box">
     </div>
   </div>
+<% elsif pr_type == 0 && pr.eligible? %>
```
Nitpick, but I think it'd be clearer if we replace `pr_type` with something like `is_bonus` and have it be a boolean?

MattIPv4 reviewed (approved):
Lgtm
